forem/forem
# Report a Message in Connect

## What type of PR is this? (check all applicable)


## Description

A user can now report abuse in Connect. Most of this work was done in #11332. I'm just closing the loop on this.

![A screenshot of the new report abuse feature in Connect](https://user-images.githubusercontent.com/833231/104380977-d6119b00-54f9-11eb-9e4f-1114a9bccee2.png)

![A screenshot of the new report abuse feature in Connect](https://user-images.githubusercontent.com/833231/104378932-b88f0200-54f6-11eb-9914-f223ba5d85ae.png)

One question I have @citizen428 since I don't know what the original specs were for this was. You can report a message, but if you try to report a second message from that user, it says that an abuse submission has already been filed. Is this how the implementation should work? I understand if they block the user as well that maybe it doesn't make sense to report another message, but when just reporting the message, I'm not sure.

## Related Tickets & Documents

This builds off the work in #11332. 

## QA Instructions, Screenshots, Recordings

1. Log in as the development mode main admin (email: admin@forem.local, password: password)
2. Make the main admin a tag moderator, for example, for the discuss tag, http://localhost:3000/admin/tags/13/edit
	1. Add the main admin as a tag moderator (their ID is 11)
	2. Click the Add Moderator button to apply the change. (This is required to be able to create new channels)
	3. Go back to Connect, localhost:3000/connect
	4. Create a channel and invite a user to the channel by clicking on the <kbd>+</kbd> icon.

![image](https://user-images.githubusercontent.com/833231/104350182-de55e000-54d1-11eb-972e-659eeb11f70b.png)

	5. The create channel modal will open. Enter a channel name and the username of another user to invite.

![image](https://user-images.githubusercontent.com/833231/104350218-e9a90b80-54d1-11eb-8efb-558137117722.png)

	6. In the newly created channel, create a message.
1. In another browser (private browser mode or a different browser), log on as the other user. All test users have a password of password. If you need to find the users email address to logon, go into the rails console and search for that user.
2. Once logged on as the other user. Accept the invitation to the new connect channel by clicking on the people icon with a badge notification.

![image](https://user-images.githubusercontent.com/833231/104350182-de55e000-54d1-11eb-972e-659eeb11f70b.png)

3. In the chat channel, hover over the message of the other user. The three dots contextual menu will appear. Hover over it to report abuse.

![image](https://user-images.githubusercontent.com/833231/104350312-05141680-54d2-11eb-881a-89cb46fcba68.png)

![image](https://user-images.githubusercontent.com/833231/104350347-0d6c5180-54d2-11eb-8ac6-ee0e1d83ba5b.png)

4. Select an abuse option. This is required to submit the form. Otherwise the report message button will stay disabled.
5. Click on the report message button. A confirmation window will notify you that the message will be reported and it will also ask you if you’d like to block the user.
	1. If you choose to block the user, the message will be reported and the user will be blocked from interacting with you on the platform.
	2. If you choose not to block the user, only the message will be reported.
6. Note, if you try to report a message from that user again, you will be notified that the report has already been submitted.

![image](https://user-images.githubusercontent.com/833231/104350389-1826e680-54d2-11eb-82e0-1e91faddd5b6.png)

### UI accessibility concerns?

The three dot contextual menu is not accessible. However, I filed is a separate issue, #12250, as this is more of a global issue in Connect.

![image](https://user-images.githubusercontent.com/833231/104382321-e591e380-54fb-11eb-8ace-d54c2a09b20f.png)

For this feature, the report abuse panel works fine with keyboard navigation and has focus styles. For the radio buttons we're using the [focus visible polyfill](https://www.npmjs.com/package/focus-visible).

![image](https://user-images.githubusercontent.com/833231/104350920-b6b34780-54d2-11eb-879a-7cab1468cece.png)

The report message button appears to be using the default browser styles. This could be something to improve upon for our button component. cc: @lisasy @ludwiczakpawel.

![image](https://user-images.githubusercontent.com/833231/104350988-cdf23500-54d2-11eb-860d-65770298e98b.png)

Also screenshare from #11332 

![Screen Recording 2020-11-09 at 11 35 AM](https://user-images.githubusercontent.com/3650216/98505988-87891b80-2280-11eb-9da0-cfdb9bbfd38b.gif)

## Added tests?

      have not been included_

## Added to documentation?

      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/)

## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](gif_link)
